### ISSUE TITLE
Add SOUND and SSP-SIR algorithms

### ISSUE THREAD

**Issue by matteo-d-m** [mne-tools/mne-denoise#32]
[ISSUE TRANSFERRED FROM MNE-PYTHON REPO by drammock on 2026-04-20]

Dear MNE-Python developers, 

I am Matteo De Matola, doctoral student at the University of Trento (Italy).

I have noticed that MNE-Python does not include an implementation of the SOUND ([Mutanen et al., 2018](https://www.sciencedirect.com/science/article/pii/S1053811917308480)) and SSP-SIR ([Mutanen et al., 2016](https://www.sciencedirect.com/science/article/pii/S1053811916301495)) algorithms. These algorithms are widely used in TMS-EEG preprocessing, but to the best of my knowledge, they are currently implemented in MATLAB only.

The absence of a reliable Python implementation for SOUND and SSP-SIR does not facilitate the adoption of MNE-Python in the TMS-EEG community. In agreement with Tuomas Mutanen, the inventor of the two algorithms, I volunteer to translate them from MATLAB to Python and integrate them into MNE. 

---

I did not open an ENH-type issue because:

1. I do not have a proposed implementation yet
2. I want to _add_ the feature, rather than request it 

If you confirm that the two algorithms are not included in MNE-Python and agree to their inclusion, I will follow up with a proposed implementation ASAP. 

Thanks,
Matteo

**Comment by PeppeDrone** [comment]
triple the need for @matteo-d-m  request! They are clearly the best, combined with the TESA pipeline.

**Comment by larsoner** [comment]
Perhaps good candidates for inclusion in https://github.com/mne-tools/mne-denoise ?

**Comment by matteo-d-m** [comment]
Yes, I think they would be!

**Comment by BabaSanfour** [comment]
@larsoner Yes, `proj="reconstruct"` is the same idea. now it only lives inside `evoked.plot()` and `plot_topomap()`, so you can see the result but not get it back as data. Promoting `_reconstruct_proj()` to public API would make it usable, not only for SSP-SIR but in general. One thing to consider though: it currently applies every projector in info across the full time range. SSP-SIR needs to reconstruct only the artifact projector it just computed, and crossfade it in over the artifact window. So we would need to add the ability to target a specific projector.

Now for the core algorithm: if I understand correctly, `_compute_mapping_matrix` is already doing the same operator we implement. It builds `surface_dots @ inv(Pᵀ·self_dots·P) @ Pᵀ`; with `self_dots ≈ L Lᵀ `and `A⁺ = Aᵀ(A Aᵀ)⁻¹`, that reduces to` L @ (P·L)⁺ `. Basically what SSP-SIR do. Maybe some numerical verification would be helpful (I can do that if we decide it's worth it to promote the function). The main difference is the model: MNE reconstructs through an implicit spherical expansion, SSP-SIR through the subject's own lead field. + truncation is hardcoded in mne; SSP-SIR wants an explicit integer rank M. So an explicit lead-field argument plus an integer rank would let both packages share one implementation.

That's two independent levels of change (Two seperate PRs). The positive is maintaining the same code only once. I don't see any negative side other than two more public functions (?) in mne.

**Comment by BabaSanfour** [comment]
@matteo-d-m @minsuzhang @PeppeDrone, we now have an initial implementation of **SSP-SIR** and **SOUND** in this PR (#46), based primarily on:

  - [Recovering TMS-evoked EEG responses masked by muscle artifacts (Mutanen et al., 2016)](https://doi.org/10.1016/j.neuroimage.2016.05.028)
  - [Automatic and robust noise suppression in EEG and MEG: The SOUND algorithm (Mutanen et al., 2018)](https://doi.org/10.1016/j.neuroimage.2017.10.021)
  - [Source-based artifact-rejection techniques for TMS–EEG (Mutanen et al., 2022)](https://doi.org/10.1016/j.jneumeth.2022.109693)

We would be very glad if you could **beta-test these implementations against your existing pipelines on real data**. Feedback on numerical behavior, parameter defaults, signal preservation, forward-model handling and performance would be great.

The PR and branch will remain open until we figure next steps for re-using mne-python functions, until we can continue the discussion here. If the PR is later merged and closed, please feel free to open issues with results, problems, suggestions, or ideas for further improvements.

 Thank you. Looking forward to your feedback!

**Comment by PeppeDrone** [comment]
Dear @BabaSanfour !

Thank you, will test for sure in the next months and come back to you with some numerical results! @matteo-d-m Did you already run a comparison with Matlab implementation?

**Comment by matteo-d-m** [comment]
Hi all,

Thank you @BabaSanfour and @snesmaeili for your work, which looks great!
I should start a new TMS-EEG project in September. That should be the chance to test your implementation and give you feedback!

**Comment by BabaSanfour** [comment]
Great news @matteo-d-m and @PeppeDrone ! by that time we will hopefully have it in mne-denoise 0.0.2.


@larsoner just pining on the message above

**Comment by larsoner** [comment]
Does the explicit lead field change the interpolation quality? We adapted the code from MNE-C and Matti has a tendency to implement things only if necessary so I'm guessing at some point he found there wasn't much difference. But that could be a separate investigation later

Yes explicit lead field or truncation / rank would be fine to add!

**Comment by BabaSanfour** [comment]
@larsoner I run a couple of more tests on this including one to test whether there is an impact on the interpolation quality. I used two different datasets ([OpenNeuro ds004024, version 1.0.1](https://openneuro.org/datasets/ds004024/versions/1.0.1) and   [MNE sample dataset](https://mne.tools/stable/generated/mne.datasets.sample.data_path.html)). I tested this by:

  1. Excluding a channel.
  2. Calculating its interpolation weights using the remaining channels.
  3. Reconstructing the signal.
  4. Comparing the reconstruction with the actual signal using correlation and normalized RMSE.
  5. Repeated for every channel:
    -  ds004024 (TMS data): 63 held-out channels per participant across three participants.
    -  MNE sample: 58 held-out channels across 288 trials.
   


* TMS-EEG dataset: the explicit lead field reduced interpolation error by around 4–6%, depending on the time window.
 * MNE sample dataset. it reduced the error by 5.7% in the baseline, 4.6% in the early response, and 3.8% in the late response. Correlation improved only a bit: 0.006–0.007 in all three windows.

So I think we have the answer on why Matti / Original C code did not use the explicit lead field, as the improvement is small. I will now merge this branch, close the issue and open a new one in mne-python.

**Comment by larsoner** [comment]
Sounds good!
